crypto/pkcs7/pk7_doit.c

```diff
@@ -55,6 +55,36 @@ ASN1_OCTET_STRING *PKCS7_get_octet_string(PKCS7 *p7)
     return NULL;
 }
 
+static ASN1_OCTET_STRING *pkcs7_get1_data(PKCS7 *p7)
+{
+    ASN1_OCTET_STRING *os = PKCS7_get_octet_string(p7);
+
+    if (os != NULL) {
+        if (os->flags & ASN1_STRING_FLAG_NDEF)
+            return os;
+        else
+            return ASN1_OCTET_STRING_dup(os);
```

Comment:
Flag `ASN1_STRING_FLAG_NDEF` is currently used by openssl-smime.
https://github.com/openssl/openssl/blob/bdba075c1962b7c69b6f40ebef9e0bf1f1325dd8/include/openssl/asn1.h.in#L139-L144

`ASN1_OCTET_STRING_dup(os)` can not be done if `ASN1_STRING_FLAG_NDEF` flag is set, because `os` does not contain any data:
os->flags = 0x0010
os->length = 0
os->data == NULL

Reply:
Hmm...does this actually fail? Or does it just end up with a non-NULL `os->data`? I've not tried it but my reading of the code suggests this might the case. I wonder if this is some deficiency in `ASN1_OCTET_STRING_dup()`? Should it really be able to handle this properly? One workaround might be to call `ASN1_OCTET_STRING_dup()` unconditionally, and then call `ASN1_STRING_set0((ASN1_STRING *)os, NULL, 0)` on the output in the `ASN1_STRING_FLAG_NDEF` case.

It seems preferable to have this function always return an allocated `ASN1_OCTET_STRING` that the caller is responsible for freeing - rather than have some obscure cases when this does not happen (especially since you named the function with `get1` which strongly implies the "caller must free" semantics)

Reply:
@olszomal has implemented calling `ASN1_OCTET_STRING_dup()` unconditionally, and it indeed simplifies calling `ASN1_OCTET_STRING_free()`.